boltstub/wiring.py

```diff
@@ -251,10 +350,50 @@ def remote_address(self):
         """
         return Address(self.__socket.getpeername())
 
+    @property
+    def _socket(self):
+        return self.__socket
+
 
 class WireError(OSError):
     """Raised when a connection error occurs."""
 
 
 class BrokenWireError(WireError):
     """Raised when a connection is broken by the network or remote peer."""
+
+
+def negotiate_socket(socket_):
+    def try_to_negotiate_websocket(socket_, buffer_):
+        encoding = "utf-8"
+        encoded = buffer_.strip().decode(encoding)
+        headers = encoded.strip().split("\r\n")
+        if "Upgrade: websocket" not in headers:
+            return False
+        for h in headers:
+            if "Sec-WebSocket-Key" in h:
+                key = h.split(" ")[1]
+        key = key + MAGIC_WS_STRING
+        encoded_key = key.encode(encoding)
+        encrypeted_key = hashlib.sha1(encoded_key).digest()
+        base64_key = base64.standard_b64encode(encrypeted_key).decode(encoding)
+        response = ("HTTP/1.1 101 Switching Protocols\r\n"
+                    + "Upgrade: websocket\r\n"
+                    + "Connection: Upgrade\r\n"
+                    + "Sec-WebSocket-Accept: %s\r\n\r\n") % (base64_key)
+        encoded_response = response.encode(encoding)
+        socket_.sendall(encoded_response)
+        return True
```

Comment:
```suggestion
    def try_to_negotiate_websocket(socket__, buffer_):
        encoding = "utf-8"
        encoded = buffer_.strip().decode(encoding)
        headers = encoded.strip().split("\r\n")
        if "Upgrade: websocket" not in headers:
            return False
        for h in headers:
            if "Sec-WebSocket-Key" in h:
                key = h.split(" ")[1]
        key = key + MAGIC_WS_STRING
        encoded_key = key.encode(encoding)
        encrypeted_key = hashlib.sha1(encoded_key).digest()
        base64_key = base64.standard_b64encode(encrypeted_key).decode(encoding)
        response = ("HTTP/1.1 101 Switching Protocols\r\n"
                    + "Upgrade: websocket\r\n"
                    + "Connection: Upgrade\r\n"
                    + "Sec-WebSocket-Accept: %s\r\n\r\n") % (base64_key)
        encoded_response = response.encode(encoding)
        socket__.sendall(encoded_response)
        return True
```

This is an optional suggestion that would avoid shadowing the outer socket_ variable.